src/core/compact_object.cc

```diff
@@ -790,8 +790,8 @@ void CompactObj::SetString(std::string_view str) {
     }
   }
 
-  if (kUseSmallStrings) {
-    if ((taglen_ == 0 && encoded.size() < (1 << 13))) {
+  if (kUseSmallStrings && SmallString::CanAllocate()) {
```

Comment:
I suggest reorder CanAllocate after comparing the  SmallString::kMaxSize. Or, alternatively, extend `CanAllocate` to pass the length parameter and check it there first